### PR TITLE
fix rpc_session bug; fix zk distributed lock bug; fix some mismatched…

### PR DESCRIPTION
**Pull request by shengofsun**
… new[]/delete to avoid unexpected problem

### CODE REVIEW

**Comment by qinzuoyan** [comment]
looks good to me, and we can resolve all dismatched new[]/delete problem by your method: "std::shared_ptr<char> buffer(new char[len], [](char* ptr){ delete []ptr; });", refer to the new issue #232.

**Comment by glglwty** [comment]
I think delete char\* and delete[] char\* is just the same, and non-trivial arrays should not be managed by shard_ptr directly.

**Comment by qinzuoyan** [comment]
@glglwty , I think though it doesn't cause any problem here, but the usage doesn't conform to the standard, and depends on the compiler implementation, and more, the valgrind memcheck tool would complain on it.

File: src/dev/cpp/utils.cpp [hunk 1]
```diff
@@ -249,7 +249,7 @@ namespace  dsn
             // optimization: zero-copy
             if (!blob.buffer_ptr())
             {
-                std::shared_ptr<char> buffer(new char[len]);
+                std::shared_ptr<char> buffer(new char[len], [](char* ptr){ delete []ptr; });
```
Comment (imzhenyu) [inline src/dev/cpp/utils.cpp:252]: is there a runtime cost for using customized deallocator if the default one already works?

**Comment by imzhenyu** [comment]
I remembered this case with valgrind, and you can add exception rules to disable the warning for this usage.

File: src/dev/cpp/utils.cpp (hunk 1 above)
Comment (glglwty) [inline src/dev/cpp/utils.cpp:252]: There won't be a runtime cost at all as long as the compiler is sane. But using `std::default_delete<T[]>` might be neater.

**Comment by shengofsun** [comment]
well, valgrind thing is not important, but why not follow the c++ standard?

**Comment by imzhenyu** [comment]
My concern is about runtime cost, but as @glglwty there should not be the case. In that case, I'm certainly glad to follow the C++ standard.

**Pull request merged**